### ISSUE TITLE
frege.java.Util has probably many bugs

### ISSUE THREAD

**Issue by Ingo60** [Frege/frege#135]
In an attempt to compare different Map implementations, I also tried  `Util.HashMap` and  found that the return types of many methods are invalid in that they don't indicate the possibiliy of `null` s returned.

See commits 49c70b852b93eaa983eb025a3135a0c393444a21 and 6b9ec9486232badd4d33b45c742453e10376f5c8 for what I fixed, but I am convinced the same will be true for many other methods.

In addition I found that when dealing with collections, it is convenient to have a `fold` function over `java.util.Iterator`s.  So added that.

Perhaps also convenient would be `foldM` and `forM_/mapM_`. The problem with Iterator.toList is that conversion to a Frege list can be  very expensive in terms of java heap size, especially because we actually need 2 lists because of the reverse. For example, to find the minimum value that is in a HashMap of size 8_000_000, it will cost you several *gigabytes* to produce the intermediate lists. Whereas 

    hm.values >>= _.iterator >>= Iterator.fold min minBound

is super fast and has no memory overhead.

This sparked the idea of making Iterator/Collection the main interfaces for Java/Frege bulk data exchange. For this, we just need to provide appropriate wrappers. A frege list `Iterator` should be a no brainer. (Not so sure about `Collection`, though). How about `Enumeration`, is this being used nowadays anymore? 

/cc @mmhelloworld

**Comment by Dierk** [comment]
Also note that there are Arrays as well (and Iterables?)
Maybe introduce an "IterableView" (analogous to ListView)?

Dierk

sent from:mobile

**Comment by seancorfield** [comment]
@Ingo60 That sounds like you would be following in Clojure's footsteps in terms of a standard "sequence" abstraction that then has specialized implementations of various "sequence functions" for each specific concrete sequence (collection) type -- and that sounds like a good idea.

**Comment by Ingo60** [comment]
@seancorfield not exactly sure about the footstep thing. My observation was that the Java collection classes heavily use the Collection and Iterator interfaces, so the low-hanging-fruit goal would be to 

  * make use of Iterators as natural as possible in Frege (though, as Iterator itself is mutable, it has to happen in the ST monad, of course)

  Because Lists, Sets, and the like are all Collections, getting an iterator for many results of java methods is already as easy as saying `>>= _.iterator`.

  * provide wrappers for Frege lists (at least) that implement Collection/Iterator interfaces, so we can easily pass lists to Java code, so they can work with them using their natural idioms (i.e. for loops, or List.new(...), etc.)

**Comment by seancorfield** [comment]
Right, I meant in terms of building to a common abstraction - which it seems is the direction you're leaning.

**Comment by mmhelloworld** [comment]
*Bump!*

> they don't indicate the possibiliy of null s returned.

This is my mistake. Sorry about that. I should have checked for nulls after generating the code.

> there should be a test suite or something that tries to check at least the less obscure classes, like collection classes, etc.

Yes, good point. This would also serve as a documentation on how to use that native module.

> This sparked the idea of making Iterator/Collection the main interfaces for Java/Frege bulk data exchange. For this, we just need to provide appropriate wrappers.
+1

**Comment by Ingo60** [comment]
The Java.Util now compiles without warnings, thanks to Frege understanding generics with bounds now.
